engines/xeen/character.cpp

```diff
@@ -0,0 +1,1840 @@
+	_itemSets[0] = &weapons;
+	_itemSets[1] = &armor;
+	_itemSets[2] = &accessories;
+	_itemSets[3] = &misc;
+}
+
+InventoryItems &InventoryItemsGroup::operator[](ItemCategory category) {
+	return *_itemSets[category];
+}
+
+void InventoryItemsGroup::breakAllItems() {
+	for (int idx = 0; idx < INV_ITEMS_TOTAL; ++idx) {
+		if ((*_itemSets[0])[idx]._id != 34) {
+			(*_itemSets[0])[idx]._bonusFlags |= ITEMFLAG_BROKEN;
+			(*_itemSets[0])[idx]._frame = 0;
+		}
+
+		(*_itemSets[1])[idx]._bonusFlags |= ITEMFLAG_BROKEN;
+		(*_itemSets[2])[idx]._bonusFlags |= ITEMFLAG_BROKEN;
+		(*_itemSets[3])[idx]._bonusFlags |= ITEMFLAG_BROKEN;
+		(*_itemSets[1])[idx]._frame = 0;
+		(*_itemSets[2])[idx]._frame = 0;
+	}
+}
+
+/*------------------------------------------------------------------------*/
+
+
+void AttributePair::synchronize(Common::Serializer &s) {
+	s.syncAsByte(_permanent);
+	s.syncAsByte(_temporary);
+}
+
+/*------------------------------------------------------------------------*/
+
+AttributePair::AttributePair() {
+	_temporary = _permanent = 0;
+}
+
+/*------------------------------------------------------------------------*/
+
+Character::Character():
+		_weapons(this), _armor(this), _accessories(this), _misc(this),
+		_items(_weapons, _armor, _accessories, _misc) {
+	clear();
+	_faceSprites = nullptr;
+	_rosterId = -1;
+}
+
+void Character::clear() {
+	_sex = MALE;
+	_race = HUMAN;
+	_xeenSide = 0;
+	_class = CLASS_KNIGHT;
+	_ACTemp = 0;
+	_birthDay = 0;
+	_tempAge = 0;
+	Common::fill(&_skills[0], &_skills[18], 0);
+	Common::fill(&_awards[0], &_awards[128], false);
+	Common::fill(&_spells[0], &_spells[39], 0);
+	_lloydMap = 0;
+	_hasSpells = false;
+	_currentSpell = 0;
+	_quickOption = QUICK_ATTACK;
+	_lloydSide = 0;
+	Common::fill(&_conditions[0], &_conditions[16], 0);
+	_townUnknown = 0;
+	_savedMazeId = 0;
+	_currentHp = 0;
+	_currentSp = 0;
+	_birthYear = 0;
+	_experience = 0;
+	_currentAdventuringSpell = 0;
+	_currentCombatSpell = 0;
+
+	_might._permanent = _might._temporary = 0;
+	_intellect._permanent = _intellect._temporary = 0;
+	_personality._permanent = _personality._temporary = 0;
+	_endurance._permanent = _endurance._temporary = 0;
+	_speed._permanent = _speed._temporary = 0;
+	_accuracy._permanent = _accuracy._temporary = 0;
+	_luck._permanent = _luck._temporary = 0;
+	_fireResistence._permanent = _fireResistence._temporary = 0;
+	_coldResistence._permanent = _coldResistence._temporary = 0;
+	_electricityResistence._permanent = _electricityResistence._temporary = 0;
+	_poisonResistence._permanent = _poisonResistence._temporary = 0;
+	_energyResistence._permanent = _energyResistence._temporary = 0;
+	_magicResistence._permanent = _magicResistence._temporary = 0;
+	_weapons.clear();
+	_armor.clear();
+	_accessories.clear();
+	_misc.clear();
+}
+
+void Character::synchronize(Common::Serializer &s) {
+	char name[16];
+	Common::fill(&name[0], &name[16], '\0');
+	strncpy(name, _name.c_str(), 16);
```

Comment:
This may lead to a buffer overflow, as strncpy doesn't take into account 0. It is safer to use Common::strlcpy which does take it into account.